pip/index.py

```diff
@@ -779,8 +791,8 @@ def _link_package_versions(self, link, search_name):
                     link, 'Python version is incorrect')
                 return
         logger.debug('Found link %s, version: %s', link, version)
-
-        return InstallationCandidate(search_name, version, link)
+        link.canonical_name = search.canonical
```

Comment:
Well, if you don't have a name, it means you're not using an index/PyPI (or are providing a direct link to a file on PyPI). Not caching it seems ok.